File: 17.电话号码的字母组合.py
#
# @lc app=leetcode.cn id=17 lang=python3
#
# [17] 电话号码的字母组合
#
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        map = {
            "2": ["a", "b", "c"],
            "3": ["d", "e", "f"],
            "4": ["g", "h", "i"],
            "5": ["j", "k", "l"],
            "6": ["m", "n", "o"],
            "7": ["p", "q", "r", "s"],
            "8": ["t", "u", "v"],
            "9": ["w", "x", "y", "z"],
        }

        def backtrace(combination: str, next_digits: str):
            if len(next_digits) != 0:
                for letter in map[next_digits[0]]:
                    backtrace(combination + letter, next_digits[1:])
            else:
                result.append(combination)

        result = []
        if len(digits) != 0:
            backtrace("", digits)
        return result

File: 17.电话号码的字母组合.cpp
/*
 * @lc app=leetcode.cn id=17 lang=cpp
 *
 * [17] 电话号码的字母组合
 */
class Solution
{
public:
    vector<string> result;
    map<char, string> phone = {{'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}};
    void traceback(string combination, string next_digits)
    {
        if (next_digits.length() != 0)
        {
            char digit = next_digits[0];
            string letters = phone[digit];
            for (int i = 0; i < letters.length(); i++)
            {
                char letter = letters[i];
                traceback(combination + letter, next_digits.substr(1));
            }
        }
        else
        {
            result.push_back(combination);
        }
    }
    vector<string> letterCombinations(string digits)
    {
        if (digits.length() != 0)
        {
            traceback("", digits);
        }
        return result;
    }
};
